### PR TITLE
Disable Load from Preset for all HW-Sources and Send Sources + disable Recall Parameter Values

### LINKED ISSUE #3035: Bitte alle HW-Sources und ihre "Snd"-Controls vom Presetsystem abkoppeln
Die vier Pedale sind bereits vom Presetsystem abgekoppelt. Das sollte für alle HW-Sources einheitlich so sein.
- [ ] Die Ribbons und die (bei Local=Off bereitgestellten) "Snd"-Controls aller HW-Sources werden mit Presets gespeichert/recallt. Das ist nicht sinnvoll.
- [ ] Damit sollten auch bei keiner der HW-Sources oder ihren "Snd"-Controls
       - in der WebUI bei Veränderung ein Originalwert angezeigt werden
       - im BOLED eine Recall-Möglichkeit über den Shift-Button angeboten werden

### PR DESCRIPTION
**Pull request by schlenkibus**
overWrite loadFromPreset for all HW-Sources and Send Sources. Recall Preset Postion is no longer supported in WebUI and Boled for those parameters, closes #3035

### CODE REVIEW

**Review by MatthiasSeeber — CHANGES_REQUESTED** [review]
A few things:
- [ ] The positions won't be recalled anymore, but the Behavior is - and when the Behavior changed, the position will be overwritten. The Behavior should also be ignored on Recall (?).
- [x] ~~(I'm on it) the Audio Engine should also not recall HW Sources anymore (see #3044)~~
- [ ] determine initial state of HW Sources (positions: first PlayController HW event?, behavior: persistent setting?)
- [ ] cleanup: The PresetMessage types (PG --> AE) should also get rid of HW Sources

File: projects/epc/playground/src/tools/IntrusiveList.h [hunk 1]
```diff
@@ -133,6 +134,15 @@ template <typename TItemPtr> class IntrusiveList
     forEach([](tItemPtr p) { delete p; });
   }
 
+  [[nodiscard]] size_t count() const
```
Comment (hhoegelo) [inline projects/epc/playground/src/tools/IntrusiveList.h:137]: Is this new method used anywhere?

File: projects/epc/playground/src/use-cases/BankUseCases.cpp [hunk 2]
```diff
@@ -264,6 +265,8 @@ Preset* BankUseCases::insertEditBufferAtPosition(int anchor)
   auto pm = getPresetManager();
   auto eb = pm->getEditBuffer();
   const auto ebWasModified = eb->isModified();
+  EditBufferStorePreparation ebsp(*eb);
```
Comment (hhoegelo) [inline projects/epc/playground/src/use-cases/BankUseCases.cpp:268]: Don't we need this in all the siblings as well, i.e. appendEditBuffer, insertEditBufferAsPresetWithUUID etc?

**Review by MatthiasSeeber — COMMENTED** [review]
Code Review:
- couldn't find anything suspicious (but to be honest, I find it very hard to review due to the complexity)

Trivial:
- acceptance tests failed (I think this is already known)
- webui has display/update problems (when for example changing a return behavior, the amounts will often not directly change their appearance)

Observations:
- Return Behavior
  - [ ] _minor_ changing a behavior will rightfully cause the preset as changed, but storing the changed preset doesn't ask for a new name
  - [x] changing a Source [NonReturn] to [ReturnZero / Center] will set the control position to 0%
  - [ ] _inconsistent_ changing a Source [Return Center] to [NonReturn] will not convert the (bi)polarity, but rather clip the value, so that any negative position will be set to 0%
  - [ ] _unknown_ I cannot observe a HW Amount being converted from "On" / 100% to 50% when changing the correspoding HW Source to [Return Center] ???
- Preset Store
  - [x] returning Sources can be in arbitrary position
- [ ] Preset Recall: several issues (see our drive document)
- Local On/Off: potential Ribbon LED update problem (see our drive document)

File: projects/epc/playground/src/playground-helpers.cpp [hunk 3]
```diff
@@ -72,6 +73,28 @@ namespace Environment
...
+  {
+    std::vector<void*> stack;
+    stack.reserve(maxFrames);
+    auto numFrames = backtrace(stack.data(), maxFrames);
+    if(numFrames == maxFrames)
+    {
+      nltools::Log::error("You could be missing some frames from your backtrace. Try to increase your maxFrames");
+    }
+
+    auto symbols = backtrace_symbols(stack.data(), numFrames); //symbols is mallocced here
```
Comment (schlenkibus) [inline projects/epc/playground/src/playground-helpers.cpp:88]: in order for symbols to be read we need to specify: -rdynamic in the c++ flags

File: projects/epc/audio-engine/src/synth/C15Synth.cpp [hunk 4]
```diff
@@ -215,11 +217,11 @@ void C15Synth::doSyncPlayground()
 
     auto sendParameterValue = m_inputEventStage.getHWSourcePositionIfLocalDisabled(hw);
     auto audioParameterValue = engineHWSourceValues[idx];
+    auto audioParameterSource = m_dsp->getValueSourceSource(hw);
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/C15Synth.cpp:220]: Consider rename into getLastValueChangeSource

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp [hunk 5]
```diff
@@ -875,8 +875,9 @@ void dsp_host_dual::onSettingTuneReference(const float _position)
   }
 }
 
-void dsp_host_dual::onSettingInitialSinglePreset()
+DSPInterface::OutputResetEventSource dsp_host_dual::onSettingInitialSinglePreset()
 {
+  // todo: rework process
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:880]: still valid comment?

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h [hunk 6]
```diff
@@ -106,6 +106,15 @@ class DSPInterface
   {
     return true;
   }
+  // ...
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h:109]: consider removing or filling comment

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h [hunk 7]
```diff
@@ -106,6 +106,15 @@ class DSPInterface
...
+  virtual void onSettingToneToggle(const uint16_t _setting)
+  {
+  }
+  virtual OutputResetEventSource onSettingInitialSinglePreset()
+  {
+    return OutputResetEventSource::None;
+  }
+  //
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h:117]: useless comment

File: projects/epc/audio-engine/src/synth/input/InputEventStage.cpp [hunk 8]
```diff
@@ -169,10 +172,23 @@ void InputEventStage::onTCDEvent()
...
     case DecoderEventType::HardwareChange:
+    {
       onHWChanged(static_cast<HardwareSource>(decoder->getKeyOrController()), decoder->getValue(), HWChangeSource::TCD,
                   false, false, false);
-
       break;
+    }
+    case DecoderEventType::PollStart:
+    {
+      // todo: disable midi send?
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/input/InputEventStage.cpp:182]: are todos still valid?

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp (hunk 5 above)
Comment (MatthiasSeeber) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.cpp:880]: Sorry in general for the occasional sloppy comment, will keep that in mind and review my code accordingly.
In this instance, yes this still is a valid comment, as there are several flaws in the whole initialization procedure. But as this topic also interferes with things like HW Polling, or for example initialized Preset Messages, a corresponding formulated task has not been made yet.

File: projects/epc/audio-engine/src/synth/input/TCDDecoder.cpp [hunk 9]
```diff
@@ -41,11 +41,35 @@ bool TCDDecoder::decode(const MidiEvent &event)
       keyOrController = channel;
       m_type = DecoderEventType::HardwareChange;
     }
-    else if(channel == 13)  //Key Pos
+    else if(_status == AE_PROTOCOL_CMD)  // PlayController Command
```
[the remark this answers is not in the document]
Comment (MatthiasSeeber) [reply projects/epc/audio-engine/src/synth/input/TCDDecoder.cpp:44]: As described, the current code reflects preparations for the HW Polling completion. I admit that the code comments are pretty verbose. I'd like to clean this up when the PlayGround implented the new HW Polling.

File: projects/epc/playground/src/parameters/HardwareSourceSendParameter.cpp [hunk 10]
```diff
@@ -208,6 +221,7 @@ void HardwareSourceSendParameter::updateIsEnabledAndSelectSiblingParameterIfAppl
       }
     }
 
+    m_sibling.onLocalEnableChanged(m_isEnabled);
```
Comment (hhoegelo) [inline projects/epc/playground/src/parameters/HardwareSourceSendParameter.cpp:224]: nitpick: pls try to be consistent in using either m_sibling or getSibl....

File: projects/epc/playground/src/parameters/Parameter.h [hunk 11]
```diff
@@ -106,8 +108,8 @@ class Parameter : public UpdateDocumentContributor,
   virtual Glib::ustring getDisplayString() const;
   virtual Glib::ustring getDisplayString(tControlPositionValue cp) const;
 
-  QuantizedValue &getValue();
-  const QuantizedValue &getValue() const;
+  virtual QuantizedValue &getValue();
```
Comment (hhoegelo) [inline projects/epc/playground/src/parameters/Parameter.h:111]: pls remove virtual

File: projects/epc/playground/src/parameters/PedalParameter.cpp [hunk 12]
```diff
@@ -42,11 +42,15 @@ void PedalParameter::undoableSetPedalMode(UNDO::Transaction *transaction, PedalM
...
     transaction->addSimpleCommand([=](UNDO::Command::State) mutable {
+       auto oldMode = m_mode;
       swapData->swapWith(m_mode);
+      auto newMode = m_mode;
+
       getValue().setScaleConverter(createScaleConverter());
       auto defValue = getDefValueAccordingToMode();
       getValue().setDefaultValue(defValue);
-      getValue().setToDefault(Initiator::INDIRECT);
+      if(oldMode != newMode && newMode != PedalModes::STAY && initiator == Initiator::EXPLICIT_USECASE)
```
Comment (hhoegelo) [inline projects/epc/playground/src/parameters/PedalParameter.cpp:52]: oldMode has to be different (already tested before)
newMode is m_mode, so you can use that

File: projects/epc/playground/src/parameters/PedalParameter.cpp [hunk 13]
```diff
@@ -285,4 +289,36 @@ bool PedalParameter::isLocalEnabled() const
...
+    }
+    else
+    {
+      auto oldSendPos = getSendParameter()->getControlPositionValue();
+      PhysicalControlParameter::setCPFromSetting(scope->getTransaction(), oldSendPos);
+    }
+  }
+  else // On -> Off
+  {
+    if(isReturning)
```
Comment (hhoegelo) [inline projects/epc/playground/src/parameters/PedalParameter.cpp:314]: consider rephrasing:

getSendParameter()->setCPFromSetting(scope->getTransaction(), getControlPositionValue());
if(isReturning)
  PhysicalControlParameter::setCPFromSetting(scope->getTransaction(), getDefValueAccordingToMode());

File: projects/epc/playground/src/playground.h [hunk 14]
```diff
@@ -81,7 +82,8 @@ enum class Initiator
   INDIRECT_SPLIT_SYNC,
   INDIRECT,
   EXPLICIT_TCD,
-  EXPLICIT_MIDI
+  EXPLICIT_MIDI,
+  INDIRECT_SETTING_CHANGED
```
Comment (hhoegelo) [inline projects/epc/playground/src/playground.h:86]: The INDIRECT_SETTING_CHANGED is actually 'write only'. Do we really need it?

File: projects/epc/playground/src/presets/EditBuffer.cpp [hunk 15]
```diff
@@ -1634,3 +1637,68 @@ Settings &EditBuffer::getSettings() const
 {
   return m_settings;
 }
+
+std::vector<double> EditBuffer::setHWSourcesToTargetPostion(UNDO::Transaction *transaction)
```
Comment (hhoegelo) [inline projects/epc/playground/src/presets/EditBuffer.cpp:1641]: Isn't his actually setHWSourcesToDefaults ?

File: projects/epc/playground/src/presets/EditBuffer.cpp [hunk 16]
```diff
@@ -1634,3 +1637,68 @@ Settings &EditBuffer::getSettings() const
...
+      {
+        auto oldMode = hw->getLastReturnModeBeforePresetLoad();
+        auto newMode = hw->getReturnMode();
+        auto oldPos = hw->getLastControlPositionValueBeforePresetLoad();
+
+        if(newMode == ReturnMode::Center && oldMode == ReturnMode::Center)
+        {
+          hw->setCPFromLoad(transaction, oldPos, false);
+        }
+        else if(newMode == ReturnMode::Center && oldMode == ReturnMode::None)
```
Comment (hhoegelo) [inline projects/epc/playground/src/presets/EditBuffer.cpp:1693]: do we need this empty branch for documentation purposes or can we just remove it?

File: projects/epc/playground/src/presets/EditBuffer.cpp [hunk 17]
```diff
@@ -1634,3 +1637,68 @@ Settings &EditBuffer::getSettings() const
...
+    {
+        hwParam->setCPFromHwui(transaction, oldPositions[idx]);
+        idx++;
+    }
+  }
+}
+
+void EditBuffer::setHWSourcesToLoadRulePostionsAndModulate(UNDO::Transaction *transaction)
+{
+  if(!Application::exists())
```
Comment (hhoegelo) [inline projects/epc/playground/src/presets/EditBuffer.cpp:1674]: Do we really need to check here? All the siblings (setHWSourcesToOldPositions ...) don't do it.

File: projects/epc/playground/src/parameters/PedalParameter.cpp [hunk 18]
```diff
@@ -42,14 +42,13 @@ void PedalParameter::undoableSetPedalMode(UNDO::Transaction *transaction, PedalM
     auto swapData = UNDO::createSwapData(mode);
 
     transaction->addSimpleCommand([=](UNDO::Command::State) mutable {
-       auto oldMode = m_mode;
+      auto oldMode = m_mode;
```
Comment (hhoegelo) [inline projects/epc/playground/src/parameters/PedalParameter.cpp:45]: oldMode seems to be unused, no?

**Pull request merged**